### PR TITLE
Displaying more then 10 bugs at once #1198

### PR DESCRIPTION
**Pull request by vipul-sharma20**
added a drop-down list to select number of bugs to be displayed at once. Added list contains options to display 10, 20, 30, 40 bugs at once.

### CODE REVIEW

**Comment by ehashman** [comment]
Hi Vipul! I skimmed through your code and tested it. Have a bit of feedback for you.

First, take a look at this screenshot. The way that this navigation element is being rendered on the page doesn't look right; this should probably be fixed.

screenshot

Second, the new `View: ###` element isn't properly aligned with the rest of the navigation elements. It would be good if it followed the same size/style so it wasn't so much further down (off by maybe 5-10px in alignment?) 

Third, you have a small fencepost error in your bug counting. The set of bugs {1..11} is actually 11 bugs because 11 is included, not 10. This holds for all the other options too. Probably want to fix that!

Last---and this is just a preference thing---you have the view # at both the top and bottom of the search bar. In my humble opinion, I prefer just having it at the top of the page. Though if anyone disagrees on that from a UX perspective, I await your feedback :)

File: mysite/search/views.py [hunk 1]
```diff
@@ -39,7 +39,6 @@
 
 # Via http://www.djangosnippets.org/snippets/1435/
 
-
```
Comment (ehashman) [inline mysite/search/views.py]: PEP8 doesn't like this change: `mysite/search/views.py:42:1: E302 expected 2 blank lines, found 1`

File: mysite/search/views.py [hunk 2]
```diff
@@ -68,7 +67,15 @@ def search_index(request, invalid_subscribe_to_alert_form=None):
 
     format = request.GET.get('format', None)
     start = int(request.GET.get('start', 1))
-    end = int(request.GET.get('end', 10))
+
+    # If user selects number of bugs from drop-down list then 'limit' is the
+    # number of bugs to display obtained by POST otherwise display 10 bugs
+    if request.method == 'POST':
+        limit =  int(request.POST.get('items'))
```
Comment (ehashman) [inline mysite/search/views.py]: PEP8 complaint: `mysite/search/views.py:74:16: E222 multiple spaces after operator`

You have a stray space before `int`.

File: mysite/search/views.py [hunk 3]
```diff
@@ -112,6 +119,9 @@ def search_index(request, invalid_subscribe_to_alert_form=None):
     next_page_query_str['start'] = end + 1
     next_page_query_str['end'] = end + diff + 1
 
+    data['items'] = limit
+    data['options'] = [10,20,30,40]
```
Comment (ehashman) [inline mysite/search/views.py]: PEP8 complaint:

```
mysite/search/views.py:123:26: E231 missing whitespace after ','
mysite/search/views.py:123:29: E231 missing whitespace after ','
mysite/search/views.py:123:32: E231 missing whitespace after ','
```

This should be `[10, 20, 30, 40]`.

File: mysite/search/templates/search/navigation.html [hunk 4]
```diff
@@ -19,14 +19,29 @@
...
 
         <span class='navigation'>
+            <span id='item-limit'>
+                <form action="/search/?{{ query_str }}" method="post">
+                {% csrf_token %}
+                View:
+                <select name="items" onchange="this.form.submit()">
+                        {% for option in options %}
+                            {% if option == items %}
+                                <option selected='selected' disabled='disabled'>{{ option }}</option>
```
Comment (ehashman) [inline mysite/search/templates/search/navigation.html]: Not sure if it's necessary to disable the selected option; -0 on this.

File: mysite/static/css/search/search.css [hunk 5]
```diff
@@ -58,14 +58,15 @@ body#search #textual_search input[type='submit'] { width: 20%; padding: 5px; }
  *---------------------- */
 body#search .results-widget { color: #555; padding: 12px; padding-bottom: 6px; float: left; width: 97%; font-size: .75em; margin-bottom: 15px; border: 0px dashed #dedede; border-bottom-width: 1px; }
     body#search .results-widget:last-child { border-bottom-width: 0; border-top-width: 1px; }
-    body#search .results-widget > span { float: left; } 
-    body#search #results-summary { width: 50%; margin-right: 2%; }
```
Comment (ehashman) [inline mysite/static/css/search/search.css]: Appreciate you getting rid of this stray whitespace, but it might have been better to do it in a separate commit to make this a little easier to review :)

**Comment by ehashman** [comment]
Thanks again for your hard work, @vipul-sharma20! And in case you haven't already, I suggest you take a look at our contributor documentation, join our devel@ mailing list and [send us a contributor statement](http://openhatch.readthedocs.org/en/latest/getting_started/handling_contributions.html#permit-us-to-share-your-work)!

**Comment by vipul-sharma20** [comment]
@ehashman Thanks for reviewing the code and sorry for all those mistakes. I will surely make the corrections suggested by you soon. For rendering the navigation element, I've used `/search/{{ query_str }}` in the `form` tag i.e. it tries to redirect to the page having query string of current page along with a POST request of the option selected from the drop-down list. That was the only way I could think of. If you can please help me with this one to make it render in a better way, it would be great !

**Comment by ehashman** [comment]
I am not sure about a better way to do this; would @paulproteus want to weigh in?

**Comment by willingc** [comment]
@vipul-sharma20 Thanks for updating the PR. Is this PR ready for review? If so, please leave a message here requesting a review. Thanks :)

**Comment by vipul-sharma20** [comment]
@willingc yes, this PR is ready for a review

File: mysite/search/templates/search/results_widget.html [hunk 6]
```diff
@@ -99,6 +99,23 @@
         <a href='#' id='expand-all-link'><img src={% version '/static/images/icons/zoom-in.png' %}>Expand all</a>
         <a href='#' id='collapse-all-link'><img src={% version '/static/images/icons/zoom-out.png' %}>Collapse all</a>
     </span>
+    <span class='active_facet_value'>
+        View:
+    </span>
+    <span id='item-limit'>
+        <form action="/search/?{{ query_str }}" method="post">
```
Comment (paulproteus) [inline mysite/search/templates/search/results_widget.html]: I believe that @vipul-sharma20 had asked for a sanity-check on this.

I believe this is pretty reasonable. One thing I'd be slightly worried about is double-escaping; if it works, though, then great.

File: mysite/search/views.py [hunk 7]
```diff
@@ -68,7 +68,15 @@ def search_index(request, invalid_subscribe_to_alert_form=None):
...
-    end = int(request.GET.get('end', 10))
+
+    # If user selects number of bugs from drop-down list then 'limit' is the
+    # number of bugs to display obtained by POST otherwise display 10 bugs
+    if request.method == 'POST':
+        limit = int(request.POST.get('items'))
+        end = start + limit - 1
+    else:
+        limit = int(request.GET.get('items', 10))
+        end = int(request.GET.get('end', limit))
```
Comment (paulproteus) [inline mysite/search/views.py]: @vipul-sharma20 I'd like to request that your code validate the `limit` so that it's always one of `[10, 20, 30, 40]`. Otherwise I am concerned people might request a `limit=100000` and then use up lots of resources on the server.

Similarly, I'd like to request that you add a unit test for this new functionality.

File: mysite/search/views.py [hunk 8]
```diff
@@ -112,6 +120,9 @@ def search_index(request, invalid_subscribe_to_alert_form=None):
     next_page_query_str['start'] = end + 1
     next_page_query_str['end'] = end + diff + 1
 
+    data['items'] = limit
+    data['options'] = [10, 20, 30, 40]
+    data['query_str'] = request.GET.urlencode()
```
Comment (paulproteus) [inline mysite/search/views.py]: This seems pretty reasonable.

File: mysite/static/css/search/search.css [hunk 9]
```diff
@@ -142,7 +143,7 @@ body#search #results ul { margin: 0; padding: 0; float: left; width: 99%; }
 
             body#search #results li .tags .project { padding: .3em .7em; float: left; }
 
-        body#search.actually_search #results form { float: left; width: 690px; }
+        body#search.actually_search #results form { float: left; width: 80px; }
```
Comment (paulproteus) [inline mysite/static/css/search/search.css]: I'm a little troubled by this particular line. The commit shrinks the width of _something_ by almost 90%, but I'm not sure why this change is needed.

**Comment by paulproteus** [comment]
My one complaint about this is that there's no horizontal space between "View" and the number e.g. `10` in the drop-down.

The easiest way (i.e., no CSS changes required) to fix that would be to adjust the drop-down so it offers these options:

* 10 per page
* 20 per page
* 30 per page
* 40 per page

and then the words "per page" are part of the drop-down, and you don't have to do any custom styling.

For now, I'm going to mark this as `needs-revision` since I think having some spacing there is essential before we show this to every visitor to the site.

**Comment by paulproteus** [comment]
@vipul-sharma20 thanks again for this pull request, and do leave a comment if you're interested in revising per the above, or if you know you aren't interested due to e.g. lack of time, or if you have questions or disagreements about the proposed revisions.
